components/ring_lights/src/effects.cpp

```diff
@@ -0,0 +1,218 @@
+	int_fast16_t diff = b - a;
+	if (diff < 0) {
+		return diff + 360;
+	}
+	return diff;
+}
+
+/* IS_BETWEEN_A_B_CLOCKWISE_DEGREES calculates whether angle c is between angle a and b
+*		in a clockwise direction
+*  @returns how far the current angle is between a and b, clockwise.
+*/
+int_fast16_t IS_BETWEEN_A_B_CLOCKWISE_DEGREES(int_fast16_t a, int_fast16_t b, int_fast16_t c) {
+	int_fast16_t relative_total_width = a + GET_CLOCKWISE_DIFF_DEGREES(a, b);
+	int_fast16_t relative_angle = a + GET_CLOCKWISE_DIFF_DEGREES(a, c);
+
+	return relative_angle - relative_total_width;
+}
+
+
+template<typename T>
+double GET_LED_ANGLE_DEGREES(T led) {
+	static_assert(std::is_arithmetic<T>::value);
+	double angle = 270 - (double) led * DEGREE_PER_LED;
+	if (angle < 0) {
+		angle = 360 + angle;
+	}
+	return angle;
+}
+
+template<typename T>
+double GET_LED_ANGLE_RAD(T led) {
+	static_assert(std::is_arithmetic<T>::value);
+	double angle = (led * RAD_PER_LED) + (RAD_PER_LED / 2) - M_PI;
+	if (angle < 0) {
+		angle = M_TAU + angle;
+	}
+	if (angle > M_PI) {
+		angle = angle - M_TAU;
+	}
+	return angle;
+}
+
+bool HSV_IS_EQUAL(hsv_t a, hsv_t b) {
+	return a.h == b.h && a.s == b.s && a.v == b.v;
+}
+
+void effects::pointer(rgb_t (& buffer)[NUM_LEDS], effect_msg& msg) {
+	int_fast16_t angle_degrees = WRAP_NEGATIVE_DEGREE(msg.param_a);
+	int_fast16_t width_degree = msg.param_b;
+	int_fast16_t width_half_pointer_degree = width_degree / 2;
+
+	hsv_t color = msg.primary_color;
+
+	for (int_fast8_t i = 0; i < NUM_LEDS; i++) {
+		double current_degree = i * DEGREE_PER_LED;
+		double degrees_to_center = GET_SMALLEST_DEGREE_DIFFERENCE(angle_degrees, current_degree);
+		double progress = 0.0;
+
+		if (degrees_to_center <= width_half_pointer_degree) {
+			progress = (degrees_to_center / width_degree) + 1;
+		}
+
+		if (0 <= progress && progress <= 1) {
+			color.value = static_cast<uint8_t>(static_cast<double>(msg.primary_color.value) * abs(progress));
+		} else if (1 <= progress && progress <= 2) {
+			color.value = static_cast<uint8_t>(static_cast<double>(msg.primary_color.value) * (1 - (progress - 1)));
+		} else {
+			color.value = 0;
+		}
+
+		buffer[i] = hsv2rgb_rainbow(color);
+		color.value = 0;
+	}
+}
+
+void effects::percent(rgb_t (& buffer)[NUM_LEDS], effect_msg& msg) {
+	int_fast16_t start = msg.param_a;
+	int_fast16_t end = msg.param_b;
+
+	// Scale `end` to percentage
+	double percent = static_cast<double>(msg.param_c) / 100;
+	int_fast16_t total_width = GET_CLOCKWISE_DIFF_DEGREES(start, end);
+	int_fast16_t correct_end = start + (total_width * (percent));
+	correct_end %= 360;
+
+	for (int_fast8_t i = 0; i < NUM_LEDS; i++) {
+		double current_degree = GET_LED_ANGLE_DEGREES(i);
+		if (IS_BETWEEN_A_B_CLOCKWISE_DEGREES(start, correct_end, current_degree)) {
+			buffer[i] = hsv2rgb_rainbow(msg.primary_color);
+		} else {
+			buffer[i] = {{0}, {0}, {0}};
+		}
+	}
+}
+
+void effects::fill(rgb_t (& buffer)[NUM_LEDS], effect_msg& msg) {
+	for (auto& i : buffer) {
```

Comment:
In this case, i doesn't really feel like a fitting name as it's mostly used in case of an index.

Reply:
changing to `led`